core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt

```diff
@@ -703,6 +703,20 @@ class UnitMovementAlgorithms(val unit: MapUnit) {
             if (unit.civInfo.isAtWarWith(firstUnit.civInfo))
                 return false
         }
+        
+        val city = tile.owningCity
+        if (tile.isWater && city !=null &&
+            !unit.civInfo.isPlayerCivilization() &&
+            unit.baseUnit.isLandUnit() &&
+            unit.civInfo.isAtWarWith(tile.getOwner()!!)) {
```

Comment:
What if the water tile is close to the city, the civ is at war with, but the tile does not belong to this city?
For example: there are 4 tiles in a line: A - B - C - D
A - an enemy city
B - the water tile of the enemy city
C - the friendly water tile with the unit to attack a city A
D - a friendly city
Both cities A and D are able to attack the water tiles B and C here.

Reply:
Also, the tiles C and D can be even an neutral water (`city == null`). Should this check work in this scenario?

Reply:
so '4' can can somehow changed to city.range

if water tile near city belongs to another city, then ai land units will die

Reply:
> if water tile near city belongs to another city or neutral waters, then ai land units will die

It means your solution does not solve the problem in full scope, but bring some additional complications, like the AI will be limited in options of capturing the coastal cities (by ships only).
Sorry, that does not sound for me as an improvement. 
Currently, AI uses quite good tactic: bring 4-6 land units, 1 of them will die on the way (that's what you trying to fix) but others will reach the land next turn successfully and will surround the city.